raft: gossiper_state_change_subscriber_proxy: on_alive: wait in background

bhalevy opened the pull request:
Wait in the error injection case in the background
so that the endpoint_lock that the gossiper holds around
this callback is freed while waiting.

Holding the gate may cause long delays in debug/dev testing
and may cause the error injection in on_alive to timeout in the
rare event that the endpoint state is not present yet in `topology_state_load`
and it ends up calling `add_saved_endpoint` which may deadlock on the
endpoint_lock held around the call to on_alive.

Fixes #16506

bhalevy commented:
Cc @tgrabiec @gusev-p

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5352/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6774/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5352/)
#### Failed Tests (243/23527):
- [auth_cluster.test_password_login_message.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.1)
- [auth_cluster.test_password_login_message.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.2)
- [auth_cluster.test_password_login_message.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.3)
- [topology.test_cluster_features.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.1)
- [topology.test_cluster_features.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.2)
- [topology.test_cluster_features.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.3)
- [topology.test_mutation_schema_change.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_mutation_schema_change_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_mutation_schema_change.debug.1)
- [topology.test_mutation_schema_change.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_mutation_schema_change_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_mutation_schema_change.debug.2)
- [topology.test_mutation_schema_change.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_mutation_schema_change_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_mutation_schema_change.debug.3)
- [topology.test_topology_remove_decom.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5352/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_topology_remove_decom_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_topology_remove_decom.debug.1)
#### Build Details:
- Duration: 10 hr
- Builder: spider8.cloudius-systems.com

tgrabiec commented on service/raft/raft_group_registry.cc:
```diff
@@ -78,12 +78,12 @@ class direct_fd_proxy : public raft::failure_detector, public direct_failure_det
     on_endpoint_change(gms::inet_address endpoint, raft::server_id id, gms::generation_type generation) {
         rslog.debug("gossiper_state_change_subscriber_proxy::on_endpoint_change() {} {} generation={}", endpoint, id, generation);
-        _address_map.add_or_update_entry(id, endpoint, generation);
+        _registry.address_map().add_or_update_entry(id, endpoint, generation);
```
> Get to the address_map in `on_endpoint_change` via the `on_endpoint_change`.

Did you mean "via raft_group_registry"?

bhalevy replied:
fixed in next version

tgrabiec commented on service/raft/raft_group_registry.cc:
```diff
@@ -377,14 +377,14 @@ future<> raft_group_registry::stop_server(raft::group_id gid, sstring reason) {
         (void)std::move(aborted)->finally([rsfg = std::move(it->second), gh = g.hold()]{});
     }
-    co_await g.close();
+    return make_ready_future();
```
Returning a ready future changes semantics of drain_on_shutdown() - it doesn't wait for servers to be stopped. drain_on_shutdown() is not only called from raft_group_registry::stop(), but also directly in main, which is problematic.

tgrabiec replied:
Not in main - in cql_test_env:

```
            // XXX: drain_on_shutdown raft before stopping the database and
            // query processor. Group registry stop raft groups
            // when stopped, and until then the groups may use
            // the database and the query processor.
            auto drain_raft = defer([this] {
                _group0_registry.invoke_on_all(&service::raft_group_registry::drain_on_shutdown).get();
            });

```

bhalevy replied:
I moved the `async_gate().stop()` call to `drain_on_shutdown()` in v2

bhalevy commented:
In v2 ([67b6eb3](https://github.com/scylladb/scylladb/commit/67b6eb39ea526c951df4008e24f2924a2647baa8)):
- fixed commitlog messages
- moved call to stop the async_gate into drain_on_shutdown

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5369/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6785/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5369/)
#### Failed Tests (7199/29758):
- [auth_cluster.test_password_login_message.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.1)
- [auth_cluster.test_password_login_message.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.2)
- [auth_cluster.test_password_login_message.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.3)
- [topology.test_cluster_features.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.1)
- [topology.test_cluster_features.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.2)
- [topology.test_cluster_features.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_cluster_features_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_cluster_features.debug.3)
- [topology.test_mutation_schema_change.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_mutation_schema_change_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_mutation_schema_change.debug.1)
- [topology.test_mutation_schema_change.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_mutation_schema_change_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_mutation_schema_change.debug.2)
- [topology.test_mutation_schema_change.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_mutation_schema_change_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_mutation_schema_change.debug.3)
- [topology.test_topology_remove_decom.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5369/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_topology_remove_decom_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_topology_remove_decom.debug.1)
#### Build Details:
- Duration: 8 hr 41 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
v3 ([168407c](https://github.com/scylladb/scylladb/commit/168407c4e91af5e0ca099116ef7674318bacbee8)):
- rebased
- make sure to close gate only once
  - made stop_servers void to reflect the fact it does everything in the background
  - and add a comment documenting that

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5384/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6801/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5384/)
#### Failed Tests (254/23581):
- [test_login_message_after_half_of_the_cluster_is_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_password_login_message/Tests___Unit_Tests___test_login_message_after_half_of_the_cluster_is_down) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_login_message_after_half_of_the_cluster_is_down)
- [test_login_message_after_half_of_the_cluster_is_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_password_login_message/Tests___Unit_Tests___test_login_message_after_half_of_the_cluster_is_down_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_login_message_after_half_of_the_cluster_is_down)
- [test_login_message_after_half_of_the_cluster_is_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_password_login_message/Tests___Unit_Tests___test_login_message_after_half_of_the_cluster_is_down_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_login_message_after_half_of_the_cluster_is_down)
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [test_rolling_upgrade_happy_path](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_rolling_upgrade_happy_path) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade_happy_path)
- [test_downgrade_after_partial_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_downgrade_after_partial_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_downgrade_after_partial_upgrade)
- [test_joining_old_node_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_joining_old_node_fails) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_joining_old_node_fails)
- [test_downgrade_after_successful_upgrade_fails](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5384/testReport/junit/%28root%29/test_cluster_features/Tests___Unit_Tests___test_downgrade_after_successful_upgrade_fails) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_downgrade_after_successful_upgrade_fails)
#### Build Details:
- Duration: 7 hr 12 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
In v4 ([35beb80](https://github.com/scylladb/scylladb/commit/35beb803e0563f1b8cfd1eb30ef534c726df4f36)):
- implemented error injection coroutine as a class member to keep it alive (as the coroutine lambda is released and loses its captures)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5387/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5387/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6803/)
#### Build Details:
- Duration: 2 hr 34 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
@tgrabiec please re-review

bhalevy commented:
@gusev-p can you please review this change?
It is a prerequisite for https://github.com/scylladb/scylladb/pull/16432
and then https://github.com/scylladb/scylladb/issues/15787

gusev-p commented:
> @gusev-p can you please review this change?
> It is a prerequisite for #16432
> and then #15787

Yes, will look at it today.

gusev-p commented on service/raft/raft_group_registry.cc:
```diff
@@ -123,13 +125,30 @@ class gossiper_state_change_subscriber_proxy: public gms::i_endpoint_state_chang
+        // wait in the background since this notification callabck is called
+        // with the endpoint locked
+        // It is safe to discard this future as the gate is closed in drain_on_shutdown()
+        (void)with_gate(_registry.async_gate(), [this, endpoint, id, generation] {
```
I think it might be better to move the `inject_with_handler` into `real_mark_alive` call site:
  * no need to complicate the matters here (`async_gate`, separate code paths for injection and non injection), simple `inject_with_handler` should continue to work in `gossiper` just like it used to work here;
  * the injection would be 'closer to the real world' there - it would imitate the network delay in getting the echo message reply from an endpoint;
  * as it would become more general, it can be reused in other tests

bhalevy replied:
makes sense

bhalevy commented:
Closing this PR in favor of https://github.com/scylladb/scylladb/pull/16432
I implemented the suggestion in https://github.com/scylladb/scylladb/pull/16507#discussion_r1438082684 as part of that series.
